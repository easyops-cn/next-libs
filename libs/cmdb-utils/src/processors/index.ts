export * from "./instance";
export * from "./object";
export * from "./formatAttrValue";
export * from "./transformConditionsToAq";
