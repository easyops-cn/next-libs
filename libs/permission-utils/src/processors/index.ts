export * from "./permissionCollection";
export * from "./permissionItem";
export * from "./processor";
