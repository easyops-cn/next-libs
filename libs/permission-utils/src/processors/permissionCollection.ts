import { PermissionItem } from "./permissionItem";
import { map, forEach, chain } from "lodash";

export class PermissionCollection {
  permissionList: any;
  constructor(instanceData: any, permissionList: any) {
    this.permissionList = map(
      permissionList,
      perm => new PermissionItem(instanceData, perm)
    );
  }

  save() {
    forEach(this.permissionList, perm => {
      perm.save();
    });
  }

  restore() {
    forEach(this.permissionList, perm => {
      perm.restore();
    });
  }

  getData() {
    const data: any = {};
    forEach(this.permissionList, perm => {
      data[perm.keyAuthorizers] = perm.isWhiteListEnabled()
        ? perm.authorizers
        : [];
    });
    return data;
  }

  getCurrentUsers() {
    return chain(this.permissionList)
      .map(perm => (perm.isWhiteListEnabled() ? perm.authorizers : []))
      .flatten()
      .uniq()
      .value();
  }
}
