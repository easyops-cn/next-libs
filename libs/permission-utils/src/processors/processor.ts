export const initPermissionOptions = (permissionList: any) => {
  return permissionList.map((perm: any) => ({
    label: perm.remark,
    value: perm.action
  }));
};
