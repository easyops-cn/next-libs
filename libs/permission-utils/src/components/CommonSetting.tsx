import React from "react";
import update from "immutability-helper";
import { DeleteOutlined, PlusOutlined } from "@ant-design/icons";
import { Table, Button, Modal, Switch, Tag } from "antd";
import * as _ from "lodash";

import { Permission, User } from "../interfaces";
import { BatchHandleUser } from "./BatchHandleUser";
import { PermissionCollection } from "../processors";
import { SelectUserOrGroup } from "./SelectUserOrGroup";
import styles from "./index.module.css";
import { CmdbModels } from "@sdk/cmdb-sdk";
export interface CommonSettingPropsDefine {
  instanceData: any;
  permissionList: Permission[];
  permissionNameOfEdit?: string;
  roleList: any;
  updateFunction?(params: Record<string, any>): void;
  stateTarget?: any;
  selectedInstances?: Record<string, any>[];
  onSingleAddUserModelOpen?(props: Record<string, any>): void;
  onSingleAddUserModelClose?(): void;
}
interface CommonSettingStateDefine {
  permissionList: Permission[];
  enableEdit: boolean;
  collections: any;
  editingKey: string;
  showAddUser: boolean;
  showBatchHandleUserModal: boolean;
  batchType: string;
  currentUsers: string[];
  temporaryUsers?: string[];
  temporaryPerms?: string[];
}

export class CommonSetting extends React.Component<
  CommonSettingPropsDefine,
  CommonSettingStateDefine
> {
  constructor(props: CommonSettingPropsDefine) {
    super(props);
    this.state = {
      permissionList: props.permissionList,
      enableEdit: false,
      collections: [],
      editingKey: "1",
      showAddUser: false,
      showBatchHandleUserModal: false,
      batchType: "add",
      currentUsers: [],
      temporaryUsers: [],
      temporaryPerms: [],
    };
  }
  componentDidMount() {
    const { instanceData, permissionList } = this.props;
    const collections = new PermissionCollection(instanceData, permissionList);
    this.setState({ collections });
    this.setState({ currentUsers: collections.getCurrentUsers() });
  }
  componentDidUpdate(prevProps: CommonSettingPropsDefine) {
    const { instanceData, permissionList, selectedInstances } = this.props;
    if (
      !_.isEqual(instanceData, prevProps.instanceData) ||
      !_.isEqual(permissionList, prevProps.permissionList)
    ) {
      const collections = new PermissionCollection(
        instanceData,
        permissionList
      );
      this.setState({ collections });
      this.setState({ currentUsers: collections.getCurrentUsers() });
    }
    if (selectedInstances !== prevProps.selectedInstances) {
      this.setState(
        {
          temporaryUsers: selectedInstances.map((instance) => instance.name),
        },
        () => {
          this.handleUsersChange();
          if (this.props.onSingleAddUserModelClose) {
            this.props.onSingleAddUserModelClose();
          }
        }
      );
    }
  }
  render() {
    const modifyButton = (
      <div className={styles.btnGroupContainer}>
        <Button onClick={this.enableEdit} id="modifyPermissionBtn">
          修改权限
        </Button>
      </div>
    );
    const buttonGroup = (
      <div className={styles.btnGroupContainer}>
        <Button
          type="primary"
          id="storeBtn"
          onClick={this.store}
          style={{ marginRight: "8px" }}
        >
          保存
        </Button>
        <Button
          id="cancelEditBtn"
          onClick={this.cancelEdit}
          style={{ marginRight: "16px" }}
        >
          取消
        </Button>
        <Button
          style={{ marginRight: "8px" }}
          id="batchAddBtn"
          icon={<PlusOutlined />}
          onClick={() => this.openBatchSetUserModel("add")}
        >
          批量添加
        </Button>
        <Button
          id="batchRemoveBtn"
          icon={<DeleteOutlined />}
          onClick={() => this.openBatchSetUserModel("remove")}
        >
          批量删除
        </Button>
      </div>
    );
    const columns = [
      {
        title: "权限名称",
        dataIndex: ["data", "remark"],
      },
      {
        title: "适用角色",
        dataIndex: ["data", "roles"],
        render: this.renderRoles,
      },
      {
        title: "启用白名单",
        dataIndex: "_whiteListEnabled",
        render: this.renderWhiteListEnabled,
      },
      {
        title: "操作",
        dataIndex: "operation",
        render: this.renderOperation,
      },
      {
        title: "用户组／用户",
        dataIndex: "authorizers",
        render: (text: any, record: any) => {
          return record._whiteListEnabled
            ? text.map((item: any) => (
                <Tag
                  color="blue"
                  key={item}
                  closable={this.state.enableEdit}
                  onClose={(e: Event) => this.removeUser(e, item, record)}
                >
                  {item}
                </Tag>
              ))
            : null;
        },
      },
    ];
    const {
      collections,
      showAddUser,
      enableEdit,
      batchType,
      showBatchHandleUserModal,
      currentUsers,
    } = this.state;
    return (
      <div className={styles.panelContainer}>
        {enableEdit ? buttonGroup : modifyButton}
        <Table columns={columns} dataSource={collections.permissionList} />
        <Modal
          destroyOnClose={true}
          title="选择用户"
          visible={showAddUser}
          onOk={this.handleStoreSingleAddUser}
          onCancel={this.closeSingleAddUserModel}
        >
          <SelectUserOrGroup handleUsersChange={this.batchHandleUserChange} />
        </Modal>
        <Modal
          destroyOnClose={true}
          title={`批量${batchType === "add" ? "添加" : "移除"}`}
          onOk={this.handleStoreBatchSetUser}
          onCancel={this.closeBatchSetUserModel}
          visible={showBatchHandleUserModal}
        >
          <BatchHandleUser
            permissionList={this.props.permissionList}
            batchType={batchType}
            currentUsers={currentUsers}
            batchHandleUserChange={this.batchHandleUserChange}
            batchHandlePermChange={this.batchHandlePermChange}
          />
        </Modal>
      </div>
    );
  }
  renderWhiteListEnabled = (text: boolean, record: any) => (
    <Switch
      defaultChecked={text}
      onChange={(checked: boolean) =>
        this.handleToggleWhiteListEnabled(checked, record)
      }
      disabled={!this.state.enableEdit}
    />
  );
  renderRoles = (text: string[]) => {
    return text.join("、");
  };
  renderOperation = (text: boolean, record: any) => {
    return record._whiteListEnabled && this.state.enableEdit ? (
      <Button
        size="small"
        onClick={() => this.openSingleAddUserModel(record)}
        id={`addUserBtn-${record.data.id}`}
        icon={<PlusOutlined />}
      />
    ) : null;
  };
  // 保存最终结果
  store = () => {
    const { collections } = this.state;
    const params: any = {};
    _.forEach(collections.permissionList, (perm) => {
      params[perm.keyAuthorizers] = perm.authorizers;
    });
    this.props.updateFunction(params);
    this.setState({ enableEdit: false });
  };
  batchHandlePermChange = (e: string[]) => {
    this.setState({ temporaryPerms: e });
  };
  batchHandleUserChange = (e: { key: string; label: string }[]) => {
    this.setState({
      temporaryUsers: e.map(
        (user: { key: string; label: string }) => user.label
      ),
    });
  };
  // 打开为单个权限添加用户弹窗
  openSingleAddUserModel = (record: any) => {
    this.batchHandlePermChange([record.data.action]);
    this.setState({
      batchType: "add",
    });
    if (this.props.onSingleAddUserModelOpen) {
      const objectIdObjectMap: Record<
        string,
        Partial<CmdbModels.ModelCmdbObject>
      > = {
        USER: {
          objectId: "USER",
          name: "用户",
        },
        USER_GROUP: {
          objectId: "USER_GROUP",
          name: "用户组",
        },
      };
      const reverseObjectIdMap: Record<string, string> = {
        USER: "USER_GROUP",
        USER_GROUP: "USER",
      };
      const onSingleAddUserModelOpen = (objectId: string) => {
        const object = objectIdObjectMap[objectId];
        const reversedObject = objectIdObjectMap[reverseObjectIdMap[objectId]];
        this.props.onSingleAddUserModelOpen({
          objectId,
          modalTitle: (
            <>
              选择{object.name}
              <Button
                type="link"
                size="small"
                onClick={() =>
                  onSingleAddUserModelOpen(reversedObject.objectId)
                }
                style={{ marginLeft: "8px" }}
              >
                切换为{reversedObject.name}
              </Button>
            </>
          ),
        });
      };
      onSingleAddUserModelOpen("USER_GROUP");
    } else {
      this.setState({ showAddUser: true });
    }
  };
  // 保存为单个权限添加用户
  handleStoreSingleAddUser = () => {
    this.handleUsersChange();
    this.setState({ showAddUser: false });
  };
  // 为单个权限删除一个用户
  removeUser = (e: Event, item: string, record: any) => {
    this.batchHandlePermChange([record.data.action]);
    this.setState({ temporaryUsers: [item], batchType: "remove" });
    this.handleUsersChange();
  };
  // 关闭单个权限添加用户弹窗
  closeSingleAddUserModel = () => {
    this.setState({ showAddUser: false });
  };
  // 打开批量添加／删除用户弹窗
  openBatchSetUserModel(batchType: string) {
    this.setState({
      batchType,
      showBatchHandleUserModal: true,
    });
  }
  // 保存批量添加／删除用户
  handleStoreBatchSetUser = () => {
    this.handleUsersChange();
    this.setState({ showBatchHandleUserModal: false });
  };
  // 关闭批量添加／删除用户弹窗
  closeBatchSetUserModel = () => {
    this.setState({ showBatchHandleUserModal: false });
  };

  handleUsersChange = () => {
    const { temporaryUsers, temporaryPerms, batchType } = this.state;
    _.forEach(this.state.collections.permissionList, (item, index) => {
      if (temporaryPerms.includes(item.data.action)) {
        if (batchType === "add") {
          item.addUsers(temporaryUsers);
        } else {
          item.removeUsers(temporaryUsers);
        }
      }
      this.setState({
        collections: update(this.state.collections, {
          permissionList: {
            [index]: { authorizers: { $set: item.authorizers } },
          },
        }),
      });
    });
    if (_.isFunction(this.state.collections.getCurrentUsers)) {
      this.setState({ currentUsers: this.state.collections.getCurrentUsers() });
    }
  };

  enableEdit = () => {
    this.setState({ enableEdit: true });
  };
  cancelEdit = () => {
    this.setState({ enableEdit: false });
  };

  handleToggleWhiteListEnabled(checked: boolean, record: any) {
    const permissionList = this.state.collections.permissionList;
    const modifiedIndex = permissionList.findIndex(
      (item: any) => item.data.id === record.data.id
    );
    permissionList[modifiedIndex]._whiteListEnabled = checked;
    if (checked && _.isEmpty(permissionList[modifiedIndex].authorizers)) {
      permissionList[modifiedIndex].authorizers.push("easyops");
    }
    this.setState((prevState) => ({
      collections: { ...prevState.collections, permissionList },
    }));
  }
}
