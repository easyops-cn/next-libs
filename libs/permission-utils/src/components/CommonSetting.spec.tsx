import React from "react";
import { shallow, mount } from "enzyme";
import { CommonSetting, CommonSettingPropsDefine } from "./CommonSetting";
import { PermissionCollection } from "../processors";
import {
  PERM_SET_OF_PACKAGE_INSTANCE,
  PERM_PACKAGE_UPDATE,
} from "../constants";
import {
  UserAdminApi_listGroupsIdName,
  UserAdminApi_ListGroupsIdNameResponseBody,
} from "@next-sdk/user-service-sdk";
import * as kit from "@next-core/brick-kit";
// const spyOnHandleHttpError = jest.spyOn(kit, "handleHttpError");
jest.mock("@next-sdk/user-service-sdk");
jest.mock("@next-core/brick-kit", () => ({
  getAuth: () => ({ username: "test" }),
}));
describe("CommonSetting", () => {
  const instanceData = {
    authUsers: "",
    cId: "1",
    category: "",
    conf: "",
    creator: "easyops",
    ctime: "2019-04-01 02:38:45",
    icon: "",
    installPath: "/tmptestpkg1",
    memo: "a",
    mtime: "2019-04-13 23:05:03",
    name: "pkg-1",
    org: "3333",
    packageId: "912cb4643e24c7723091509fa9ad82c4",
    platform: "linux",
    repoId: "",
    repoPath: "/3333/91/2c/b4/643e24c7723091509fa9ad82c4",
    source: "",
    style: "",
    type: "1",
    deleteAuthorizers: ["easyops", "06031"],
    readAuthorizers: ["easyops", "06032"],
    updateAuthorizers: ["easyops", "06034", "group1"],
    testAuthorizers: ["000"],
  };
  const permissionList = [
    {
      action: "deploy:package_delete",
      id: "5b7b87a27e5a2e22bb29bcbf",
      remark: "包删除",
      resource: {
        system: "deploy",
        condition: {
          deleteAuthorizers: "%user",
          name: "package",
          system: "deploy",
        },
      },
      roles: [
        "系统管理员",
        "应用运维",
        "应用测试",
        "应用开发",
        "test",
        "测试角色-没有easyops+系统管理菜单",
      ],
      system: "持续部署",
      user: [
        "willniu123",
        "easyops",
        "qw",
        "qwe",
        "ea",
        "willniu",
        "lightjiao",
        "user3",
      ],
      _actionWeight: 1,
      _category: "a-0",
    },
    {
      action: "deploy:package_update",
      id: "5b7b87a27e5a2e22bb29bcc0",
      remark: "包编辑",
      resource: {
        system: "deploy",
        condition: {
          updateAuthorizers: "%user",
          name: "package",
          system: "deploy",
        },
      },
      roles: [
        "系统管理员",
        "应用运维",
        "应用测试",
        "应用开发",
        "test",
        "测试角色-没有easyops+系统管理菜单",
      ],
      system: "持续部署",
      user: [
        "willniu123",
        "easyops",
        "qw",
        "qwe",
        "ea",
        "willniu",
        "lightjiao",
        "user3",
      ],
      _actionWeight: 2,
      _category: "a-0",
    },
    {
      action: "deploy:package_read",
      id: "5b7b87a27e5a2e22ba29bcc4",
      remark: "包查看",
      resource: {
        system: "deploy",
        condition: {
          readAuthorizers: "%user",
          name: "package",
          system: "deploy",
        },
      },
      roles: [
        "系统管理员",
        "应用运维",
        "应用测试",
        "应用开发",
        "test",
        "测试角色-没有easyops+系统管理菜单",
      ],
      system: "持续部署",
      user: [
        "willniu123",
        "easyops",
        "qw",
        "qwe",
        "ea",
        "willniu",
        "lightjiao",
        "user3",
      ],
      _actionWeight: 100,
      _category: "a-0",
    },
    {
      action: "deploy:package_test",
      id: "5b7b87a27e5a2e22ba29bcc9",
      remark: "包测试",
      resource: {
        system: "deploy",
        condition: {
          testAuthorizers: "%user",
          name: "package",
          system: "deploy",
        },
      },
      roles: [
        "系统管理员",
        "应用运维",
        "应用测试",
        "应用开发",
        "test",
        "测试角色-没有easyops+系统管理菜单",
      ],
      system: "持续部署",
      user: ["data"],
      _actionWeight: 100,
      _category: "a-0",
    },
  ];
  const roleList = [
    {
      permission: [
        "5b7b87737e5a2e22b829bcbd",
        "5b7b87737e5a2e22ba29bcac",
        "5b7b87737e5a2e22b929bcac",
      ],
      role: "系统管理员",
      user: ["easyops", "user3", "group1", "lightjiao"],
      org: 3333,
      id: "5b7b876e7e5a2e22b829bcac",
    },
    {
      permission: ["5c1373157e5a2e9d26040ac3", "5c1373157e5a2e9d24040aa8"],
      role: "应用运维",
      user: ["willniu123", "willniu"],
      org: 3333,
      id: "5b7b876e7e5a2e22b829bcb6",
    },
    {
      permission: [
        "5b7b87797e5a2e22b829bcc0",
        "5b7b87a27e5a2e22ba29bcb9",
        "5b7b87a27e5a2e22ba29bcba",
      ],
      role: "应用测试",
      user: [],
      org: 3333,
      id: "5b7b876e7e5a2e22b829bcb7",
    },
    {
      permission: [
        "5b7b87797e5a2e22b829bcc0",
        "5b7b87a27e5a2e22ba29bcb9",
        "5b7b87a27e5a2e22ba29bcba",
      ],
      role: "应用开发",
      user: [],
      org: 3333,
      id: "5b7b876e7e5a2e22b829bcb8",
    },
    {
      permission: [
        "5b7b87a27e5a2e22ba29bcc4",
        "5b7b87a27e5a2e22bb29bcbe",
        "5b7b87a27e5a2e22bb29bcc0",
      ],
      role: "test",
      user: ["qwe", "group2"],
      org: 3333,
      id: "5c3ee1b37e5a2e3497d23eba",
    },
    {
      permission: [
        "5b7b87797e5a2e22b829bcc0",
        "5b7b877d7e5a2e22b829bcc2",
        "5b7b877d7e5a2e22bb29bcb1",
      ],
      role: "测试角色-没有easyops+系统管理菜单",
      user: ["qw"],
      org: 3333,
      id: "5c10c30d7e5a2e021395c091",
    },
  ];
  const userIdToName = {
    alanname123: "alanname123",
    anntest: "anntest",
    autodiscover: "autodiscover",
    autotool: "autotool",
    lightjiao: "lightjiao",
    luoke01: "luoke01",
    mannyzheng: "这个昵称是用来检测如果内容过于冗长以致于文本换行会是什么效果",
    openapi: "openapi",
    qw: "qw",
    qwe: "qwe",
    reolei: "reolei",
    rfid: "rfid",
    robertma: "robertma",
    shawn: "shawn",
    shijian: "shijian",
    szkingdom: "szkingdom",
    test110: "test110",
    unisims: "unisims",
    user1: "user1",
    user2: "user2",
    user3: "user3",
    user4: "user4",
    willniu: "willniu",
    willniu123: "willniu123",
    yasuo: "yasuo",
    zachary: "zachary",
    zhangxing: "zhangxing",
  };
  const userGroupIdToName = {
    group1: "group1",
    group2: "group2",
    group3: "group3",
    group4: "group4",
  };
  const permissionSet = PERM_SET_OF_PACKAGE_INSTANCE;
  const permissionNameOfEdit = PERM_PACKAGE_UPDATE;
  const updateFunction = jest.fn();
  const onSingleAddUserModelClose = jest.fn();
  const stateTarget = {};
  const props: CommonSettingPropsDefine = {
    instanceData,
    // permissionSet,
    permissionNameOfEdit,
    permissionList,
    roleList,
    // userIdToName,
    // userGroupIdToName,
    updateFunction,
    stateTarget,
  };
  const wrapper = shallow(<CommonSetting {...props} />);
  const instance = wrapper.instance() as CommonSetting;
  it("should work", () => {
    expect(wrapper).toBeTruthy();
  });
  it("componentDidUpdate", async () => {
    const spy = UserAdminApi_listGroupsIdName as jest.Mock;
    spy.mockResolvedValue({} as UserAdminApi_ListGroupsIdNameResponseBody);
    spy.mockRejectedValue("error");
    wrapper.setProps({
      selectedInstances: [{ instanceId: "fake_id", name: "fake_name" }],
    });
    expect(instance.state.temporaryUsers).toEqual(["fake_name"]);
  });

  it("should enable editing when click the modify button", () => {
    wrapper.find("#modifyPermissionBtn").simulate("click");
    expect(instance.state.enableEdit).toEqual(true);
  });
  it("should disable editing when click the modify button", () => {
    wrapper.find("#cancelEditBtn").simulate("click");
    expect(instance.state.enableEdit).toEqual(false);
  });
  it("should call store function and exit editing status when click the store button", () => {
    const mockProps = { ...props, updateFunction: jest.fn() };
    const mockWrapper = shallow(<CommonSetting {...mockProps} />);
    const mockInstance = mockWrapper.instance() as CommonSetting;
    mockInstance.enableEdit();
    const spy = jest.spyOn(mockProps, "updateFunction");
    mockWrapper.find("#storeBtn").simulate("click");
    expect(spy).toHaveBeenCalled();
    expect(mockInstance.state.enableEdit).toEqual(false);
  });
  it("should open the batch add user modal when click the batch add button", () => {
    instance.enableEdit();
    const spy = jest.spyOn(instance, "openBatchSetUserModel");
    wrapper.find("#batchAddBtn").simulate("click");
    expect(spy).toHaveBeenCalled();
    expect(instance.state.showBatchHandleUserModal).toEqual(true);
  });
  it("should join roles when render", () => {
    expect(instance.renderRoles(["系统管理员", "测试角色"])).toEqual(
      "系统管理员、测试角色"
    );
  });
  it("should work roles when renderOption", () => {
    const record = instance.state.collections.permissionList[0];
    expect(instance.renderOperation(true, record)).toBeTruthy();
  });
  it("should work when enable a whiteList", () => {
    const record = instance.state.collections.permissionList[0];
    instance.handleToggleWhiteListEnabled(true, record);
    expect(instance.state.collections.permissionList[0]).toEqual(record);
    wrapper.setProps({
      fillCurrentLoginUser: true,
    });
    const record1 = instance.state.collections.permissionList[3];
    instance.handleToggleWhiteListEnabled(true, record1);
    expect(instance.state.collections.permissionList[3]).toEqual(record1);
  });
  it("should work when batchHandlePermChange is called", () => {
    instance.batchHandlePermChange(["deploy:package_read"]);
    expect(instance.state.temporaryPerms).toEqual(["deploy:package_read"]);
  });
  it("should work when batchHandleUserChange is called", () => {
    instance.setState({
      idMapName: { ":789": "ai" },
    });
    instance.batchHandleUserChange([
      { label: "easyops_111", key: "12345", value: "12345" },
      { label: "ai", key: "789", value: "789" },
    ]);
    expect(instance.state.temporaryUsers).toEqual(["easyops_111", ":789"]);
  });
  it("should work when batch add users", () => {
    instance.setState({
      batchType: "add",
      temporaryPerms: [
        "deploy:package_read",
        "deploy:package_update",
        "deploy:package_delete",
      ],
      temporaryUsers: ["anntest"],
    });
    instance.handleUsersChange();
    const permissionList = instance.state.collections.permissionList;
    const deleteItem = permissionList.find(
      (item: any) => item.keyAuthorizers === "deleteAuthorizers"
    );
    expect(deleteItem.authorizers).toEqual(["easyops", "06031", "anntest"]);
  });
  it("should work when batch remove users", () => {
    instance.setState({
      batchType: "remove",
      temporaryPerms: [
        "deploy:package_read",
        "deploy:package_update",
        "deploy:package_delete",
      ],
      temporaryUsers: ["06031"],
    });
    instance.handleUsersChange();
    const permissionList = instance.state.collections.permissionList;
    const deleteItem = permissionList.find(
      (item: any) => item.keyAuthorizers === "deleteAuthorizers"
    );
    expect(deleteItem.authorizers).toEqual(["easyops", "anntest"]);
  });
  it("should work when close a whiteList", () => {
    const record = instance.state.collections.permissionList[0];
    instance.handleToggleWhiteListEnabled(false, record);
    expect(instance.state.collections.permissionList[0].authorizers).toEqual(
      []
    );
  });
});
