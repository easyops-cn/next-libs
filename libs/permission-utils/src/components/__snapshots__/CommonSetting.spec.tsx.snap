// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CommonSetting should work 1`] = `
<div
  className="panelContainer"
>
  <div
    className="btnGroupContainer"
  >
    <Button
      block={false}
      ghost={false}
      htmlType="button"
      id="modifyPermissionBtn"
      loading={false}
      onClick={[Function]}
    >
      修改权限
    </Button>
  </div>
  <Table
    columns={
      Array [
        Object {
          "dataIndex": Array [
            "data",
            "remark",
          ],
          "title": "权限名称",
        },
        Object {
          "dataIndex": Array [
            "data",
            "roles",
          ],
          "render": [Function],
          "title": "适用角色",
        },
        Object {
          "dataIndex": "_whiteListEnabled",
          "render": [Function],
          "title": "启用白名单",
        },
        Object {
          "dataIndex": "operation",
          "render": [Function],
          "title": "操作",
        },
        Object {
          "dataIndex": "authorizers",
          "render": [Function],
          "title": "用户组／用户",
        },
      ]
    }
    dataSource={
      Array [
        PermissionItem {
          "_whiteListEnabled": true,
          "authorizers": Array [
            "easyops",
            "06031",
          ],
          "data": Object {
            "_actionWeight": 1,
            "_category": "a-0",
            "action": "deploy:package_delete",
            "id": "5b7b87a27e5a2e22bb29bcbf",
            "remark": "包删除",
            "resource": Object {
              "condition": Object {
                "deleteAuthorizers": "%user",
                "name": "package",
                "system": "deploy",
              },
              "system": "deploy",
            },
            "roles": Array [
              "系统管理员",
              "应用运维",
              "应用测试",
              "应用开发",
              "test",
              "测试角色-没有easyops+系统管理菜单",
            ],
            "system": "持续部署",
            "user": Array [
              "willniu123",
              "easyops",
              "qw",
              "qwe",
              "ea",
              "willniu",
              "lightjiao",
              "user3",
            ],
          },
          "instanceData": Object {
            "authUsers": "",
            "cId": "1",
            "category": "",
            "conf": "",
            "creator": "easyops",
            "ctime": "2019-04-01 02:38:45",
            "deleteAuthorizers": Array [
              "easyops",
              "06031",
            ],
            "icon": "",
            "installPath": "/tmptestpkg1",
            "memo": "a",
            "mtime": "2019-04-13 23:05:03",
            "name": "pkg-1",
            "org": "3333",
            "packageId": "912cb4643e24c7723091509fa9ad82c4",
            "platform": "linux",
            "readAuthorizers": Array [
              "easyops",
              "06032",
            ],
            "repoId": "",
            "repoPath": "/3333/91/2c/b4/643e24c7723091509fa9ad82c4",
            "source": "",
            "style": "",
            "type": "1",
            "updateAuthorizers": Array [
              "easyops",
              "06034",
              "group1",
            ],
          },
          "keyAuthorizers": "deleteAuthorizers",
        },
        PermissionItem {
          "_whiteListEnabled": true,
          "authorizers": Array [
            "easyops",
            "06034",
            "group1",
          ],
          "data": Object {
            "_actionWeight": 2,
            "_category": "a-0",
            "action": "deploy:package_update",
            "id": "5b7b87a27e5a2e22bb29bcc0",
            "remark": "包编辑",
            "resource": Object {
              "condition": Object {
                "name": "package",
                "system": "deploy",
                "updateAuthorizers": "%user",
              },
              "system": "deploy",
            },
            "roles": Array [
              "系统管理员",
              "应用运维",
              "应用测试",
              "应用开发",
              "test",
              "测试角色-没有easyops+系统管理菜单",
            ],
            "system": "持续部署",
            "user": Array [
              "willniu123",
              "easyops",
              "qw",
              "qwe",
              "ea",
              "willniu",
              "lightjiao",
              "user3",
            ],
          },
          "instanceData": Object {
            "authUsers": "",
            "cId": "1",
            "category": "",
            "conf": "",
            "creator": "easyops",
            "ctime": "2019-04-01 02:38:45",
            "deleteAuthorizers": Array [
              "easyops",
              "06031",
            ],
            "icon": "",
            "installPath": "/tmptestpkg1",
            "memo": "a",
            "mtime": "2019-04-13 23:05:03",
            "name": "pkg-1",
            "org": "3333",
            "packageId": "912cb4643e24c7723091509fa9ad82c4",
            "platform": "linux",
            "readAuthorizers": Array [
              "easyops",
              "06032",
            ],
            "repoId": "",
            "repoPath": "/3333/91/2c/b4/643e24c7723091509fa9ad82c4",
            "source": "",
            "style": "",
            "type": "1",
            "updateAuthorizers": Array [
              "easyops",
              "06034",
              "group1",
            ],
          },
          "keyAuthorizers": "updateAuthorizers",
        },
        PermissionItem {
          "_whiteListEnabled": true,
          "authorizers": Array [
            "easyops",
            "06032",
          ],
          "data": Object {
            "_actionWeight": 100,
            "_category": "a-0",
            "action": "deploy:package_read",
            "id": "5b7b87a27e5a2e22ba29bcc4",
            "remark": "包查看",
            "resource": Object {
              "condition": Object {
                "name": "package",
                "readAuthorizers": "%user",
                "system": "deploy",
              },
              "system": "deploy",
            },
            "roles": Array [
              "系统管理员",
              "应用运维",
              "应用测试",
              "应用开发",
              "test",
              "测试角色-没有easyops+系统管理菜单",
            ],
            "system": "持续部署",
            "user": Array [
              "willniu123",
              "easyops",
              "qw",
              "qwe",
              "ea",
              "willniu",
              "lightjiao",
              "user3",
            ],
          },
          "instanceData": Object {
            "authUsers": "",
            "cId": "1",
            "category": "",
            "conf": "",
            "creator": "easyops",
            "ctime": "2019-04-01 02:38:45",
            "deleteAuthorizers": Array [
              "easyops",
              "06031",
            ],
            "icon": "",
            "installPath": "/tmptestpkg1",
            "memo": "a",
            "mtime": "2019-04-13 23:05:03",
            "name": "pkg-1",
            "org": "3333",
            "packageId": "912cb4643e24c7723091509fa9ad82c4",
            "platform": "linux",
            "readAuthorizers": Array [
              "easyops",
              "06032",
            ],
            "repoId": "",
            "repoPath": "/3333/91/2c/b4/643e24c7723091509fa9ad82c4",
            "source": "",
            "style": "",
            "type": "1",
            "updateAuthorizers": Array [
              "easyops",
              "06034",
              "group1",
            ],
          },
          "keyAuthorizers": "readAuthorizers",
        },
      ]
    }
    rowKey="key"
  />
  <Modal
    confirmLoading={false}
    destroyOnClose={true}
    maskTransitionName="fade"
    okType="primary"
    onCancel={[Function]}
    onOk={[Function]}
    title="选择用户"
    transitionName="zoom"
    visible={false}
    width={520}
  >
    <SelectUserOrGroup
      handleUsersChange={[Function]}
    />
  </Modal>
  <Modal
    confirmLoading={false}
    destroyOnClose={true}
    maskTransitionName="fade"
    okType="primary"
    onCancel={[Function]}
    onOk={[Function]}
    title="批量添加"
    transitionName="zoom"
    visible={false}
    width={520}
  >
    <BatchHandleUser
      batchHandlePermChange={[Function]}
      batchHandleUserChange={[Function]}
      batchType="add"
      currentUsers={
        Array [
          "easyops",
          "06031",
          "06034",
          "group1",
          "06032",
        ]
      }
      permissionList={
        Array [
          Object {
            "_actionWeight": 1,
            "_category": "a-0",
            "action": "deploy:package_delete",
            "id": "5b7b87a27e5a2e22bb29bcbf",
            "remark": "包删除",
            "resource": Object {
              "condition": Object {
                "deleteAuthorizers": "%user",
                "name": "package",
                "system": "deploy",
              },
              "system": "deploy",
            },
            "roles": Array [
              "系统管理员",
              "应用运维",
              "应用测试",
              "应用开发",
              "test",
              "测试角色-没有easyops+系统管理菜单",
            ],
            "system": "持续部署",
            "user": Array [
              "willniu123",
              "easyops",
              "qw",
              "qwe",
              "ea",
              "willniu",
              "lightjiao",
              "user3",
            ],
          },
          Object {
            "_actionWeight": 2,
            "_category": "a-0",
            "action": "deploy:package_update",
            "id": "5b7b87a27e5a2e22bb29bcc0",
            "remark": "包编辑",
            "resource": Object {
              "condition": Object {
                "name": "package",
                "system": "deploy",
                "updateAuthorizers": "%user",
              },
              "system": "deploy",
            },
            "roles": Array [
              "系统管理员",
              "应用运维",
              "应用测试",
              "应用开发",
              "test",
              "测试角色-没有easyops+系统管理菜单",
            ],
            "system": "持续部署",
            "user": Array [
              "willniu123",
              "easyops",
              "qw",
              "qwe",
              "ea",
              "willniu",
              "lightjiao",
              "user3",
            ],
          },
          Object {
            "_actionWeight": 100,
            "_category": "a-0",
            "action": "deploy:package_read",
            "id": "5b7b87a27e5a2e22ba29bcc4",
            "remark": "包查看",
            "resource": Object {
              "condition": Object {
                "name": "package",
                "readAuthorizers": "%user",
                "system": "deploy",
              },
              "system": "deploy",
            },
            "roles": Array [
              "系统管理员",
              "应用运维",
              "应用测试",
              "应用开发",
              "test",
              "测试角色-没有easyops+系统管理菜单",
            ],
            "system": "持续部署",
            "user": Array [
              "willniu123",
              "easyops",
              "qw",
              "qwe",
              "ea",
              "willniu",
              "lightjiao",
              "user3",
            ],
          },
        ]
      }
    />
  </Modal>
</div>
`;
